python/tank/context.py

```diff
@@ -668,6 +668,141 @@ def create_empty(tk):
     """
     return Context(tk)
 
+def from_entity_dictionary(tk, entity_dictionary):
+    """
+    """
+    if not isinstance(entity_dictionary, dict):
+        raise TankError("Cannot create a context from an empty or invalid entity dictionary!")
+    
+    if "type" not in entity_dictionary:
+        raise TankError("Cannot create a context without an entity type!")
+    
+    if "id" not in entity_dictionary:
+        raise TankError("Cannot create a context without an entity id!")
+    
+    # prep our context data structure
+    context = {
+        "tk": tk,
+        "project": None,
+        "entity": None,
+        "step": None,
+        "user": None,
+        "task": None,
+        "additional_entities": []
+    }
+
+    entity_type = entity_dictionary["type"]
+    entity_id = entity_dictionary["id"]
+
+    # try to determine the various entities from the entity dictionary:
+    project = entity = step = task = None
+    fallback_to_ctx_from_entity = False
+    if entity_type == "Project":
+        # find entities for a project context
+        project = entity_dictionary
+    elif entity_type == "Task":
+        # find entities for a task context
+        task = entity_dictionary
+        if "project" not in task or "entity" not in task or "step" not in task:
+            fallback_to_ctx_from_entity = True
+        else:
+            project = task["project"]
+            entity = task["entity"]
+            step = task["step"]
+    elif entity_type in ["PublishedFile", "TankPublishedFile"]:
+        # special case handling for published files:
+        if "task" in entity_dictionary:
+            # construct a task context
+            return from_entity_dictionary(tk, entity_dictionary["task"])
+        elif "entity" in entity_dictionary:
+            # construct an entity context
+            return from_entity_dictionary(tk, entity_dictionary["entity"])
+        elif "project" in entity_dictionary:
+            # construct project context
+            return from_entity_dictionary(tk, entity_dictionary["project"])
+        else:
```

Comment:
Is it possible for task to be in the dict but have a None value?  If it is, this would error... but would it be better to just cascade down to the next case?